File: cli/src/history.ts
import { readFileSync, writeFileSync, existsSync, mkdirSync, readdirSync } from "node:fs";
import { join } from "node:path";
import { homedir } from "node:os";
import type { Decision } from "./decisions.js";

const HISTORY_DIR = join(homedir(), ".defer", "history");

function ensureHistoryDir(): void {
  if (!existsSync(HISTORY_DIR)) {
    mkdirSync(HISTORY_DIR, { recursive: true });
  }
}

export interface HistoryEntry {
  task: string;
  decisions: Decision[];
  completedAt: string;
  cost: number;
  duration: number;
}

/** Convert a task description into a filesystem-safe slug. */
function slugify(text: string): string {
  return text
    .toLowerCase()
    .replace(/[^a-z0-9]+/g, "-")
    .replace(/^-|-$/g, "")
    .slice(0, 50);
}

/** Save a completed session to history. */
export function saveToHistory(
  task: string,
  decisions: Decision[],
  cost: number,
  duration: number
): string {
  ensureHistoryDir();

  const timestamp = new Date().toISOString().replace(/[:.]/g, "-");
  const slug = slugify(task);
  const filename = `${timestamp}-${slug}.json`;

  const entry: HistoryEntry = {
    task,
    decisions,
    completedAt: new Date().toISOString(),
    cost,
    duration,
  };

  writeFileSync(join(HISTORY_DIR, filename), JSON.stringify(entry, null, 2));
  return filename;
}

/** List recent history entries, newest first. */
export function listHistory(limit = 20): string[] {
  ensureHistoryDir();
  return readdirSync(HISTORY_DIR)
    .filter((f) => f.endsWith(".json"))
    .sort()
    .reverse()
    .slice(0, limit);
}

/** Load a specific history entry by filename. Returns null if not found. */
export function loadHistoryEntry(filename: string): HistoryEntry | null {
  const path = join(HISTORY_DIR, filename);
  if (!existsSync(path)) return null;
  try {
    return JSON.parse(readFileSync(path, "utf-8"));
  } catch {
    return null;
  }
}

File: go/internal/history/history.go
package history

import (
	"encoding/json"
	"fmt"
	"os"
	"path/filepath"
	"regexp"
	"sort"
	"strings"
	"time"

	"github.com/defer-ai/cli/internal/decision"
)

// Entry represents a completed session.
type Entry struct {
	Task        string              `json:"task"`
	Decisions   []decision.Decision `json:"decisions"`
	CompletedAt string              `json:"completedAt"`
	Duration    int64               `json:"duration"` // ms
}

func historyDir() (string, error) {
	home, err := os.UserHomeDir()
	if err != nil {
		return "", err
	}
	dir := filepath.Join(home, ".defer", "history")
	return dir, os.MkdirAll(dir, 0o755)
}

var slugRe = regexp.MustCompile(`[^a-z0-9]+`)

func slugify(text string) string {
	s := slugRe.ReplaceAllString(strings.ToLower(text), "-")
	s = strings.Trim(s, "-")
	if len(s) > 50 {
		s = s[:50]
	}
	return s
}

// Save persists a completed session.
func Save(task string, decisions []decision.Decision, duration int64) (string, error) {
	dir, err := historyDir()
	if err != nil {
		return "", err
	}

	ts := strings.ReplaceAll(strings.ReplaceAll(time.Now().UTC().Format(time.RFC3339), ":", "-"), ".", "-")
	filename := fmt.Sprintf("%s-%s.json", ts, slugify(task))

	entry := Entry{
		Task:        task,
		Decisions:   decisions,
		CompletedAt: time.Now().UTC().Format(time.RFC3339),
		Duration:    duration,
	}

	data, err := json.MarshalIndent(entry, "", "  ")
	if err != nil {
		return "", err
	}
	return filename, os.WriteFile(filepath.Join(dir, filename), data, 0o644)
}

// List returns recent history filenames, newest first.
func List(limit int) ([]string, error) {
	dir, err := historyDir()
	if err != nil {
		return nil, err
	}
	entries, err := os.ReadDir(dir)
	if err != nil {
		return nil, err
	}
	var files []string
	for _, e := range entries {
		if !e.IsDir() && strings.HasSuffix(e.Name(), ".json") {
			files = append(files, e.Name())
		}
	}
	sort.Sort(sort.Reverse(sort.StringSlice(files)))
	if limit > 0 && len(files) > limit {
		files = files[:limit]
	}
	return files, nil
}

// Load reads a specific history entry.
func Load(filename string) (*Entry, error) {
	dir, err := historyDir()
	if err != nil {
		return nil, err
	}
	data, err := os.ReadFile(filepath.Join(dir, filename))
	if err != nil {
		return nil, err
	}
	var entry Entry
	return &entry, json.Unmarshal(data, &entry)
}
